### PR TITLE
feat: embedded runner

### PR DESCRIPTION
**Pull request by larme**
## What does this PR address?

Fixes #(issue)

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

**Comment by larme** [comment]
`serve_grpc_development` and `serve_http_development` in `bentoml/serve.py`, `grpc_dev_api_server.py` and `http_dev_api_server.py` is not needed because we are merging development and production servers. I kept them now for easier reference when changing test cases. Will clean them up when tests are ready.

File: src/bentoml/serve.py [hunk 1]
```diff
@@ -319,7 +324,12 @@ def serve_http_production(
     from ._internal.configuration.containers import BentoMLContainer
 
     working_dir = os.path.realpath(os.path.expanduser(working_dir))
-    svc = load(bento_identifier, working_dir=working_dir, standalone_load=True)
+
+    standalone_load: bool = False if development_mode else True
```
Comment (sauyon) [inline src/bentoml/serve.py:328]: Why are we loading standalone if in development mode?

File: src/bentoml/serve.py [hunk 2]
```diff
@@ -500,12 +527,41 @@ def serve_http_production(
...
+
+            # NOTE: {} is faster than dict()
+            plugins = [
+                # reloader plugin
+                {
+                    "use": "bentoml._internal.utils.circus.watchfilesplugin.ServiceReloaderPlugin",
+                    "working_dir": working_dir,
+                    "bentoml_home": bentoml_home,
+                },
+            ]
```
Comment (sauyon) [inline src/bentoml/serve.py:553]: Let's allow people to specify reload outside of development mode, I think that's fine.

**Review by sauyon — COMMENTED** [review]
Looks good, but a few questions.

Do you think it would be doable to pass the embedded runners through the batching layer? Probably for another PR.

File: src/bentoml/serve.py (hunk 1 above)
Comment (larme) [reply src/bentoml/serve.py:211]: The old development code load in standalone.

File: src/bentoml/serve.py (hunk 2 above)
Comment (larme) [reply src/bentoml/serve.py]: I have test reload for production server and seems working ok. So now I apply `--reload` for both production and development.

File: src/bentoml/serve.py (hunk 1 above)
Comment (sauyon) [reply src/bentoml/serve.py:211]: Hm, @parano @aarnphm do you remember why this was?

File: src/bentoml/serve.py [hunk 3]
```diff
@@ -884,11 +955,13 @@ def serve_grpc_production(
     )
     scheme = "https" if BentoMLContainer.ssl.enabled.get() else "http"
 
+    close_child_stdin: bool = False if development_mode else True
```
[the remark this answers is not in the document]
Comment (larme) [reply src/bentoml/serve.py:661]: This is also the way in old codes. I guess in development we should allow stdin access?

File: src/bentoml/_internal/models/model.py [hunk 4]
```diff
@@ -342,13 +344,22 @@ def to_runner(
...
         from ..runner import Runner
 
+        # TODO: @larme @yetone run this branch only yatai version is incompatible with embedded runner
+        yatai_version = os.environ.get("YATAI_T_VERSION")
+        if embedded and yatai_version:
+            logger.warning(
+                f"Yatai of version {yatai_version} is incompatible with embedded runner, set `embedded=False` for runner {name}"
+            )
+            embedded = False
+
```
Comment (larme) [inline src/bentoml/_internal/models/model.py:354]: This is for old version of yatai, which will ignore `BentoRunnerInfo`'s embedded field and spawn a pod anyway. In this case, we will still use non-embedded mode for embedded runner.

When yatai is compatible with embedded field, let's add a version check here. @yetone

**Comment by larme** [comment]
@sauyon please review again, thanks.

File: src/bentoml/_internal/models/model.py (hunk 4 above)
Comment (yetone) [reply src/bentoml/_internal/models/model.py:354]: LGTM!

**Comment by larme** [comment]
will put docs and examples changes in another pr

**Review by yetone — APPROVED** [review]

File: src/bentoml/serve.py (hunk 1 above)
Comment (aarnphm) [reply src/bentoml/serve.py:211]: I believe this is a hack around the import module issue. I don't think standalone_load makes sense

File: src/bentoml/serve.py [hunk 5]
```diff
@@ -32,9 +34,9 @@
 
 SCRIPT_RUNNER = "bentoml_cli.worker.runner"
 SCRIPT_API_SERVER = "bentoml_cli.worker.http_api_server"
-SCRIPT_DEV_API_SERVER = "bentoml_cli.worker.http_dev_api_server"
+SCRIPT_DEV_API_SERVER = "bentoml_cli.worker.http_api_server"
 SCRIPT_GRPC_API_SERVER = "bentoml_cli.worker.grpc_api_server"
-SCRIPT_GRPC_DEV_API_SERVER = "bentoml_cli.worker.grpc_dev_api_server"
+SCRIPT_GRPC_DEV_API_SERVER = "bentoml_cli.worker.grpc_api_server"
```
Comment (aarnphm) [inline src/bentoml/serve.py:39]: remove this since it is not relevant anymore.

File: src/bentoml/serve.py [hunk 6]
```diff
@@ -968,11 +750,44 @@ def serve_grpc_production(
...
+                    "use": "bentoml._internal.utils.circus.watchfilesplugin.ServiceReloaderPlugin",
+                    "working_dir": working_dir,
+                    "bentoml_home": bentoml_home,
+                },
+            ]
+
+        arbiter_kwargs["plugins"] = plugins
+        arbiter_kwargs["debug"] = True if sys.platform != "win32" else False
+        arbiter_kwargs["loggerconfig"] = SERVER_LOGGING_CONFIG
+        arbiter_kwargs["loglevel"] = "WARNING"
```
Comment (aarnphm) [inline src/bentoml/serve.py:785]: It seems like we are having this branch the same twice for both grpc and HTTP serve function, should we refactor out to a shared function?

**Review by aarnphm — COMMENTED** [review]
final comments, and then it is gtg

File: src/bentoml/serve.py [hunk 7]
```diff
@@ -171,123 +173,6 @@ def find_triton_binary():
     return binary
 
 
-@inject
-def serve_http_development(
```
Comment (aarnphm) [inline src/bentoml/serve.py:175]: Technically this is a breaking change if users are using this function. I think we should deprecate it until 1.1

I'm ok with making this breaking in 1.1

File: src/bentoml/serve.py (hunk 6 above)
Comment (larme) [reply src/bentoml/serve.py]: done

File: src/bentoml/serve.py (hunk 7 above)
Comment (larme) [reply src/bentoml/serve.py:175]: is this a public api we document?

File: src/bentoml_cli/serve.py [hunk 8]
```diff
@@ -20,13 +20,21 @@ def add_serve_command(cli: click.Group) -> None:
...
         show_default=True,
     )
+    @click.option(
+        "--production",
+        type=click.BOOL,
+        help="Deprecated option, by default BentoServer will run in production mode",
+        is_flag=True,
+        default=True,
+        show_default=True,
+    )
```
Comment (larme) [inline src/bentoml_cli/serve.py:37]: @aarnphm just added `--production` back for backward compatibility.

File: src/bentoml/serve.py (hunk 1 above)
Comment (larme) [reply src/bentoml/serve.py]: after some code reading, I think `standalone_load=True` should be only used by runner/api server, so I remove this part

File: src/bentoml/server.py [hunk 9]
```diff
@@ -185,7 +185,7 @@ def __init__(
         self,
         bento: str | Bento | Tag,
         reload: bool = False,
-        production: bool = False,
+        production: bool = True,
```
Comment (larme) [inline src/bentoml/server.py:188]: @aarnphm I'm updating HTTPServer and GrpcServer's production default to True, to match our cli default

File: src/bentoml/serve.py (hunk 7 above)
Comment (aarnphm) [reply src/bentoml/serve.py:175]: we don't, but i believe there are users that use this internal API

File: src/bentoml/server.py (hunk 9 above)
Comment (aarnphm) [reply src/bentoml/server.py:188]: sg

**Review by aarnphm — COMMENTED** [review]
Overal LGTM, just one more thing is that maybe we can deprecate `serve_http_development` and just call `serve_http_production` with api_worker=1 underneath, so that it won't break user code.

**Comment by aarnphm** [comment]
Also some of the tests are failing. I can take over this since I have a tests PR.

**Comment by larme** [comment]
> Also some of the tests are failing. I can take over this since I have a tests PR.

the tests failed because we switch BentoServer's production default from False to True. The production server need quite longer to shutdown hence I bumped the timeout from 10 to 60.

**Comment by larme** [comment]
@aarnphm please take a look again, thanks!

**Review by aarnphm — APPROVED** [review]
LGTM. Since gRPC is not yet fully released, I think it is good to break `serve_grpc_development`.

**Pull request merged**